trackintel/preprocessing/triplegs.py

```diff
@@ -200,124 +155,167 @@ def generate_trips(stps_input, tpls_input, gap_threshold=15, id_offset=0, print_
+    ls = []
+    for key, values in trip2spt_map.items():
+        for value in values:
+            ls.append([value, key])
+    temp = pd.DataFrame(ls, columns=['sptsIndex', 'trip_id'])
+    spts = spts.merge(temp, how='left', left_on='id', right_on='sptsIndex')
+    spts.drop(columns='sptsIndex', inplace=True)
+    spts['id'] = ids
+    spts.set_index('id', inplace=True)
+    
+    # assign prev_trip_id to spts
+    ids = spts.index
+    spts = spts.merge(trips[['id', 'destination_staypoint_id']], 
+                    how ='left', 
+                    left_on ='id', 
+                    right_on='destination_staypoint_id')
+    spts.rename(columns={"id":"prev_trip_id"}, inplace=True)
+    spts.drop(columns={"destination_staypoint_id"}, inplace=True)
+    
+    # assign next_trip_id to spts
+    # restore for merging
+    spts['id'] = ids
+    spts.set_index('id', inplace=True)
+    spts = spts.merge(trips[['id', 'origin_staypoint_id']], 
+                      how ='left', 
+                      left_on ='id', 
+                      right_on='origin_staypoint_id')
+    spts.rename(columns={"id":"next_trip_id"}, inplace=True)
+    spts.drop(columns={"origin_staypoint_id"}, inplace=True)
+    # set back the ids as index
+    spts['id'] = ids
+    spts.set_index('id', inplace=True)
+    
+    tpls.drop(['type'], axis=1, inplace=True)
+    spts.drop(['type'], axis=1, inplace=True)
+    
+    trips.drop(['tpls', 'spts'], axis=1, inplace=True)
+    trips.set_index('id', inplace=True)
+    return spts, tpls, trips
 
-    for user_id_this in spts_tpls['user_id'].unique():
-        unknown_activity = {'user_id': user_id_this, 'activity': True, 'id': np.nan}
-
-        spts_tpls_this = spts_tpls[spts_tpls['user_id'] == user_id_this]
-        # assert (spts_tpls_this['started_at'].is_monotonic)  # this is expensive and should be replaced
-
-        origin_activity = unknown_activity
-        temp_trip_stack = []
-        before_first_trip = True
-        in_trip = False
-
-        for _, row in spts_tpls_this.iterrows():
-            if print_progress:
-                if trip_id_counter % 100 == 0:
-                    if not trip_id_counter in dont_print_list:
-                        print("trip number: {}".format(trip_id_counter))
-                        dont_print_list.append(trip_id_counter)
-
-            # check if we can start a new trip
-            # (we make sure that we start the trip with the most recent activity)
-            if in_trip is False:
-                # If there are several activities in a row, we skip until the last one
-                if row['activity'] and row['activity_next']:
-                    continue
-
-                # if this is the last activity before the trip starts, reset the origin
-                elif row['activity']:
+def _generate_trips_user(df, gap_threshold):
+
+    unknown_activity = {'user_id': df['user_id'].unique()[0], 'activity': True, 'id': np.nan}
```

Comment:
I think the `df['user_id'].unique()[0]` is a bit confusing. I am ware that we enter this function after grouping by user id, so couldn't we simply tyke `df['user_id'].iloc[0]` ?
On the other way, we could use the `df['user_id'].unique()[0]` for an assert that could help the readability. 
```python
user_id = df['user_id'].unique()
assert len(user_id) == 1
user_id = user_id[0]
unknown_activity = {'user_id': user_id, 'activity': True, 'id': np.nan}
```

Reply:
Replaced and solved